### ISSUE TITLE
Linking error when completing C++ SDK build

### ISSUE THREAD

**Issue by Afectr** [aws/aws-sdk-cpp#1269]
### What platform/OS are you using?
Amazon Linux/g++ (GCC) 7.2.1 20170915

### Which version of the SDK?
1.6.0 (I think)

### What compiler are you using? what version?
AWS C9

### What are your CMake arguments?
Linking CXX executable aws-cpp-sdk-core-tests

### Can you provide a TRACE level log? (sanitize any sensitive information)
/bin/sh: line 1:  5833 Segmentation fault      ./aws-cpp-sdk-core-tests --gtest_shuffle --gtest_repeat=3
make[2]: *** [aws-cpp-sdk-core-tests/aws-cpp-sdk-core-tests] Error 139
make[2]: *** Deleting file `aws-cpp-sdk-core-tests/aws-cpp-sdk-core-tests'
make[1]: *** [aws-cpp-sdk-core-tests/CMakeFiles/aws-cpp-sdk-core-tests.dir/all] Error 2
make: *** [all] Error 2

https://docs.aws.amazon.com/cloud9/latest/user-guide/sample-cplusplus.html#sample-cplusplus-sdk
Hi, I am following the c++ sample and trying to build c++ SDK into my environment via AWS c9. I am getting the above errors as I come to make completion via building SDK. How do I resolve my build? Maybe there is a folder/directory issue and a file cannot be found? This is my first issue here so I hope I am not breaking any guidelines posting this.

**Comment by KaibaLopez** [comment]
Hi @Afectr, 
I just followed the tutorial and was not able to replicate your error, so I'll need a bit more of info.
When you created the environment on cloud9, what type was it? (t2.micro - m4.large)
What were the cmake arguments you used?
Amazon linux environment by default comes with gcc 4, can you verify that you have gcc and g++ 7 or higher installed, and that the build process used that?

**Comment by Afectr** [comment]
Hi @KaibaLopez, I think I started with t2.micro, and then I switched to the 8GB ram one t2.large as my environment kept running out of memory.
For cmake arguments I used the inputs from the C++ sample to build all without arguments.
I have upgraded to g++ 7 as my build stopped in midway previously. g++ (GCC) 7.2.1 20170915 (Red Hat 7.2.1-2)

**Comment by KaibaLopez** [comment]
I forgot to ask this in the previous comment but, I'm guessing this is failing for you before you get to the sample code section in the tutorial?

Segmentation fault here is kind of weird, and it works on my cloud9 instance, so just to be sure I'm going to ask you to do (probably repeat) these steps:
```
sudo yum -y install libcurl-devel openssl-devel libuuid-devel cmake3
git clone https://github.com/aws/aws-sdk-cpp.git
cd aws-sdk-cpp
mkdir sdk_build
cd sdk_build
cmake3 .. -DBUILD_ONLY=s3
make
```
If it fails, please provide the full log of the build.

**Comment by Afectr** [comment]
Hi, this worked @KaibaLopez. Not sure what I missed the first time but this fixed it. Thanks

**Comment by Afectr** [comment]
Hi, I took a break from this project and got stuck on the next part of linking.

`Paths:
/home/ec2-user/environment/aws-sdk-cpp/    #git source from here
/home/ec2-user/environment/sdk_build/    #build DIR`

`Commands:
mkdir sdk_build && cd sdk_build
cmake ../aws-sdk-cpp -DBUILD_ONLY="s3"
sudo make && sudo make install    #not sure if make install was required
#created below files
cmake3 -Daws-sdk-cpp_DIR=sdk_build .
sudo make
Run: /home/ec2-user/environment/sdk_build/s3-demo.cpp my-test-bucket us-east-2
`

`Files:
/home/ec2-user/environment/sdk_build/s3-demo.cpp
/home/ec2-user/environment/sdk_build/CMakeLists.txt
#both created from https://docs.aws.amazon.com/cloud9/latest/user-guide/sample-cplusplus.html#sample-cplusplus-sdk-code`

Outcome:
Getting undefined reference errors (about 20 odd of them)
`Aws::Client::ClientConfiguration::ClientConfiguration()`
`undefined reference to `Aws::S3::S3Client::~S3Client()`

During Install configuration step of sudo make install I see files installed to `/usr/local/lib64/cmake/AWSSDK`
`/usr/local/include/aws`

I tried target_include _libraries command with arguments like `/usr/local/include/aws/core/` /usr/local/include/aws/s3/model/ /usr/local/lib64/aws-c-common/`, but I am probably using these incorrectly.

**Comment by KaibaLopez** [comment]
Hi @Afectr,
When you did:
```
cmake3 -Daws-sdk-cpp_DIR=sdk_build .
sudo make
```
I think it didn't find the libraries, can you try the absolute path there and tell me what you get?

**Comment by Afectr** [comment]
Hi @KaibaLopez, thanks for your response. This is the output I got:

`admin:~/environment $ cd sdk_build`
`admin:~/environment/sdk_build $ cmake3 -Daws-sdk-cpp_DIR=/home/ec2-user/environment/sdk_build . `                                                   
-- TARGET_ARCH not specified; inferring host OS to be platform compilation target
-- Building AWS libraries as shared objects
-- Generating linux build config
-- Building project version: 1.7.252
-- Configuring done
-- Generating done
-- Build files have been written to: /home/ec2-user/environment/sdk_build/.deps
[ 33%] Built target AwsCCommon
[ 66%] Built target AwsChecksums
[100%] Built target AwsCEventStream
-- Third-party dependencies are installed at: /home/ec2-user/environment/sdk_build/.deps/install
--   Zlib library: /usr/lib64/libz.so
-- Encryption: Openssl
--   Openssl include directory: /usr/include
--   Openssl library: /usr/lib64/libssl.so;/usr/lib64/libcrypto.so
-- Http client: Curl
--   Curl include directory: /usr/include
--   Curl library: /usr/lib64/libcurl.so
-- Considering s3
-- Updating version info to 1.7.252
-- Custom memory management enabled; stl objects now using custom allocators
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    aws-sdk-cpp_DIR


-- Build files have been written to: /home/ec2-user/environment/sdk_build
`admin:~/environment/sdk_build $ sudo make`
[ 20%] Built target aws-cpp-sdk-core
[ 89%] Built target aws-cpp-sdk-s3
[ 91%] Built target testing-resources
[ 91%] Built target aws-cpp-sdk-s3-integration-tests
[100%] Built target aws-cpp-sdk-core-tests


PS: Sorry about the formatting. I do not know how to put it into code parameters

**Comment by KaibaLopez** [comment]
so, a couple things I'd like you to try here.
First try and change the command in the c++ Runner to call the s3-demo instad of s3-demo.cpp, so: 
`/home/ec2-user/environment/sdk_build/s3-demo my-test-bucket us-east-2`
And Second, try running it from the console instead of the C++ runner, so in the console, while in the sdk_build directory try:
`./s3-demo my-test-bucket us-east-2`
Let me know if either of those worked for you.

**Comment by Afectr** [comment]
Hi @KaibaLopez, I tried those commands you mentioned but the issue the executable is missing even after compilation via the cmake commands. I do not have 's3-demo' anywhere in my directories. I think something went wrong when going through the CMakeLists.txt.

```
Running /home/ec2-user/environment/sdk_build/s3-demo
g++: error: /home/ec2-user/environment/sdk_build/s3-demo: No such file or directory
g++: fatal error: no input files
compilation terminated.
```
```
admin:~/environment $ ./s3-demo my-test-bucket us-east-2
bash: ./s3-demo: No such file or directory
```

**Comment by KaibaLopez** [comment]
Ah I see, 
so when you are doing "cmake3 -Daws-sdk-cpp_DIR=/home/ec2-user/environment/sdk_build ." it seems like it's actually building the sdk instead of s3-demo.cpp...
try this, move s3-demo.cpp and CMakeLists.txt (the ones the tutorial is telling you to make) into another directory, preferably root (~/enviroment), and from that directory run both:
```
cmake3 -Daws-sdk-cpp_DIR=/home/ec2-user/environment/sdk_build .
make
```
if that fails, it may be an issue with the CMakeLists.txt, in which case, try changing it to:
```
cmake_minimum_required(VERSION 3.5)
project(s3-demo)
find_package(AWSSDK REQUIRED COMPONENTS s3)
add_executable(s3-demo s3-demo.cpp)
target_link_libraries(s3-demo ${AWSSDK_LINK_LIBRARIES})
```

**Comment by Afectr** [comment]
Hi @KaibaLopez,

I have tried your suggestions the above and I got the following outputs:

My environment (CMakeLists.txt and s3-demo.cpp in /home/ec2-user/environment/):
```
aws-sdk-cpp       #/home/ec2-user/environment
- .c9
- aws-sdk-cpp
- CMakeFiles
- sdk_build
-- .deps
-- aws-cpp-sdk-core
-- aws-cpp-sdk-core-tests
-- aws-cpp-sdk-s3
-- aws-cpp-sdk-s3-integration-tests
-- AWSSDK
- CMakeLists.txt
- s3-demo.cpp
```

First suggestion:
Input:
```
cmake3 -Daws-sdk-cpp_DIR=/home/ec2-user/environment/sdk_build .
make
```
Output:
```
CMake Warning at /usr/share/cmake3/Modules/CMakeFindDependencyMacro.cmake:65 (find_package):
  By not providing "Findaws-c-event-stream.cmake" in CMAKE_MODULE_PATH this
  project has asked CMake to find a package configuration file provided by
  "aws-c-event-stream", but CMake did not find one.

  Could not find a package configuration file provided by
  "aws-c-event-stream" with any of the following names:

    aws-c-event-streamConfig.cmake
    aws-c-event-stream-config.cmake

  Add the installation prefix of "aws-c-event-stream" to CMAKE_PREFIX_PATH or
  set "aws-c-event-stream_DIR" to a directory containing one of the above
  files.  If "aws-c-event-stream" provides a separate development package or
  SDK, be sure it has been installed.
Call Stack (most recent call first):
  sdk_build/aws-cpp-sdk-core/aws-cpp-sdk-core-config.cmake:2 (find_dependency)
  sdk_build/aws-sdk-cpp-config.cmake:1 (include)
  CMakeLists.txt:14 (find_package)


CMake Warning at CMakeLists.txt:14 (find_package):
  Found package configuration file:

    /home/ec2-user/environment/sdk_build/aws-sdk-cpp-config.cmake

  but it set aws-sdk-cpp_FOUND to FALSE so package "aws-sdk-cpp" is
  considered to be NOT FOUND.  Reason given by package:

  aws-sdk-cpp could not be found because dependency aws-c-event-stream could
  not be found.



-- Configuring done
-- Generating done
-- Build files have been written to: /home/ec2-user/environment
```

Second suggestion (Changing CMakeLists.txt):
Input:
```
cmake3 -Daws-sdk-cpp_DIR=/home/ec2-user/environment/sdk_build .
make
```
Ouput:
```
CMake Error at CMakeLists.txt:3 (find_package):
  By not providing "FindAWSSD.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "AWSSD", but
  CMake did not find one.

  Could not find a package configuration file provided by "AWSSD" with any of
  the following names:

    AWSSDConfig.cmake
    awssd-config.cmake

  Add the installation prefix of "AWSSD" to CMAKE_PREFIX_PATH or set
  "AWSSD_DIR" to a directory containing one of the above files.  If "AWSSD"
  provides a separate development package or SDK, be sure it has been
  installed.


-- Configuring incomplete, errors occurred!
See also "/home/ec2-user/environment/CMakeFiles/CMakeOutput.log".
See also "/home/ec2-user/environment/CMakeFiles/CMakeError.log".
```

**Comment by KaibaLopez** [comment]
Ok,
it looks like you got a typo there
```
Could not find a package configuration file provided by "AWSSD" with any of the following names:
```
Can you verify that?

**Comment by KaibaLopez** [comment]
hmmm I'm not sure how to replicate this issue , from my side it all runs...
So I can only suggest you to kind of start over clean, delete all the folders except .c9 and do these steps again:
```
cd ~
git clone https://github.com/aws/aws-sdk-cpp.git
cd aws-sdk-cpp
mkdir sdk_build
cd sdk_build
cmake3 .. -DBUILD_ONLY=s3
make install
cd ~
*create CMakeLists.txt and s3-demo.cpp here (outside the aws-sdk-cpp folder)*
cmake3 . 
make
```
After that you should have an executable file called s3-demo which, you'll run with the runner created on step 3 of the tutorial (not the default c++ runner).
Hopefully this works, sorry it's been a bit of a mess to go through this.

**Comment by Afectr** [comment]
Hi @KaibaLopez, I have made it work on my end. Here were what I found out (and the solutions).
- AWS SDK looks for install path of sdk_build, so add into ~/environment/CMakeLists.txt the following:
`set(AWSSDK_ROOT_DIR, "/usr/local/")`
- The project builds aws sdk as shared libs, but ~/environment/CMakeLists.txt looks for static libs for some reason. Fix:
`set(BUILD_SHARED_LIBS ON)`
```
cmake_minimum_required(VERSION 3.5)
project(s3-demo)
set(AWSSDK_ROOT_DIR, "/usr/local/")
set(BUILD_SHARED_LIBS ON)
find_package(AWSSDK REQUIRED COMPONENTS s3)
add_executable(s3-demo s3-demo.cpp)
target_link_libraries(s3-demo ${AWSSDK_LINK_LIBRARIES})
```
Built with commands above and tested in terminal with
` ./s3-demo my-test-bucket us-west-2 `
